### ISSUE TITLE
RFE: support a LARGE feature level without also doing DNSSEC

### ISSUE THREAD

**Issue by yrro** [systemd/systemd#10672]
**systemd version the issue has been seen with**
> 239

**Used distribution**
 > Debian

**Expected behaviour you didn't see**
 > `resolvectl query ad.domain.example` success

**Unexpected behaviour you saw**
 > `$ resolvectl query ad.domain.example`
 > `ad.domain.example: resolve call failed: Access denied`

**Steps to reproduce the problem**
The label in question has enough A records that the response won't fit into the 512-byte limit of a classical DNS UDP datagram. In addition, the server does not support DNS-over-TCP.

I have examined what's going on with Wireshark and have observed the following sequence of packets:

> initial query
1. UDP query A ad.domain.example OPT (EDNS0, length 4096, DAU+DHU+N3U)
2. UDP query A ad.domain.example OPT (EDNS0, length 4096, DAU+DHU+N3U)
3. UDP response to no.1, reply code: Format error
4. UDP response to no.2, reply code: Format error
> resolved retries with a smaller buffer size
5. UDP query A ad.domain.example OPT (EDNS0, length 512, DAU+DHU+N3U)
6. UDP query A ad.domain.example OPT (EDNS0, length 512, DAU+DHU+N3U)
7. UDP response to no.5, reply code: Format error
8. UDP response to no.6, reply code: Format error
> resolved retries without dnssec and still with the smaller buffer size
9. UDP query A ad.domain.example OPT (EDNS0, length 512, no DNSSEC)
10. UDP query A ad.domain.example OPT (EDNS0, length 512, no DNSSEC)
11. UDP response to no.9 with truncated bit set
> resolved switches to TCP
12. TCP SYN to DNS server
13. UDP response to no.10 with truncated bit set
14. TCP SYN to DNS server
15. TCP RST, ACK in response to no.12
16. TCP RST, ACK in response to no.14

I can provide a dump of the above exchange privately as long as the recipient keeps it confidential.

I have noticed that I see the same FORMERR response when querying the server with `dig @server ad.domain.example`. In this case, dig does use EDNS, but does not include the DAU+DHU+N3U options (and I do not know how to make it do so); however it does by default use the COOKIE option. If I add `+nocookie` to the command line then EDNS is still used, with no options, and the query succeeds!

My hypothesis is that the server (which is actually a load-balancing appliance) responds with FORMERR if a query contains any EDNS options that it does not recognize.

If I add `+bufsize=512 ` as well then I'm reproducing the queries sent by resolved in packet nos.9 and 10, which rightly fails because the response would be too large.

So, what is the actual bug in resolved? Well, it is that after it gives up on DNSSEC, it keeps the 512-byte buffer size. It would work if an additional layer of fallback was added, without DNSSEC but with the 4096 byte buffer size.

I tried setting `DNSSEC=no` in `resolved.conf` but this did not help: in this case resolved does not even try to request a 4096-byte response in the first place.

**Comment by hexchain** [comment]
Encountered (possibly) the same issue today. The domain `alibabacloud.com` has records that cannot fit within 512B (at least for some people & me, YMMV), and I have DNSSEC turned off due to #9867.

`dig`, `drill` and `kdig` either have EDNS0 turned on by default or retry with TCP when the answer is truncated, but `systemd-resolved` does not:

```
$ resolvectl query alibabacloud.com
alibabacloud.com: resolve call failed: 'alibabacloud.com' does not have any RR of the requested type
```

And after switching to a server that supports DNSSEC:
```
$ sudo resolvectl dns wlp58s0 1.1.1.1 8.8.8.8
$ sudo resolvectl dnssec wlp58s0 yes
$ sudo resolvectl reset-server-features  # otherwise the above line does not have any effect
$ resolvectl status wlp58s0
Link 4 (wlp58s0)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
DefaultRoute setting: yes
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: yes
    DNSSEC supported: yes
  Current DNS Server: 1.1.1.1
         DNS Servers: 1.1.1.1
                      8.8.8.8
          DNS Domain: ~.
                      default.local
```
```
$ resolvectl query alibabacloud.com
alibabacloud.com: 47.88.198.17                 -- link: wlp58s0
                  47.88.198.24                 -- link: wlp58s0
                  47.88.198.31                 -- link: wlp58s0
                  47.88.198.34                 -- link: wlp58s0
                  47.88.198.35                 -- link: wlp58s0
                  47.88.198.36                 -- link: wlp58s0
                  47.88.198.37                 -- link: wlp58s0
                  47.88.198.38                 -- link: wlp58s0
                  47.88.198.39                 -- link: wlp58s0
                  47.88.198.40                 -- link: wlp58s0
                  47.88.251.161                -- link: wlp58s0
                  47.88.251.162                -- link: wlp58s0
                  47.88.251.163                -- link: wlp58s0
                  47.88.251.164                -- link: wlp58s0
                  47.88.251.165                -- link: wlp58s0
                  47.88.251.166                -- link: wlp58s0
                  47.88.251.168                -- link: wlp58s0
                  47.88.251.169                -- link: wlp58s0
                  47.88.251.170                -- link: wlp58s0
                  47.88.251.171                -- link: wlp58s0
                  47.88.251.172                -- link: wlp58s0
                  47.88.251.173                -- link: wlp58s0
                  47.88.251.174                -- link: wlp58s0
                  47.88.251.175                -- link: wlp58s0
                  47.88.251.176                -- link: wlp58s0
                  47.88.251.177                -- link: wlp58s0
                  47.88.251.178                -- link: wlp58s0
                  47.88.251.179                -- link: wlp58s0
                  47.88.251.180                -- link: wlp58s0
                  47.88.128.163                -- link: wlp58s0

-- Information acquired via protocol DNS in 60.5ms.
-- Data is authenticated: no
```

**Comment by poettering** [comment]
Is this still an issue? I tried to reproduce the issue here, but afaics we correctly fall back to TCP if the TC bit is set in the responses we get. 

Is this reproducible on current systemd-resolved versions?

**Comment by yrro** [comment]
The bug isn't that the fallback to TCP failed. That was never a problem except that at the time I filed this bug, my work DNS servers did not respond to DNS queries via TCP.

If I configure iptables to drop outgoing DNS queries via TCP then I still see this behaviour with version 246.6. That is, resolved does not attempt a non-DNSSEC EDNS query over UDP with the 'UDP Payload size' field set to 4096 bytes.

What I would like is for step 9 to set the 'UDP Payload size' field to 4096 bytes instead of 512.

**Comment by poettering** [comment]
So far we only bother with the large payload size if DNSSEC is on, since typically only then responses are that large. For non-DNSSEC the responses are usually much shorter, and hence we don't bother, knowing that there's always TCP if the short size doesn't work.

**Comment by poettering** [comment]
So, as I understand due to manual firewalling you had a server that was reachable via UDP, but not via TCP, and you requested some large non-DNSSEC response, and that hence failed?

I am not sure this is really a bug though. I mean, your case is a very special case of misconfiguration, no? EDNS0 is much newer than DNS-via-TCP. Hence it sounds strange to try the modern stuff if the common baseline doesn't even work?

**Comment by yrro** [comment]
The manual firewalling is necessary for me to recreate the conditions that I into in the wild when I originally filed this bug. Namely, when I started at my current employer DNS was UDP only. TCP packets to the DNS servers were dropped. And if your DNS/UDP queries used DNSSec options then the servers responded with `FORMERR`.

And we have a lot of domain controllers, so even something as simple as looking for `A` RRs for `ad.domain.example.com` failed, because `resolved` would not even try to use EDNS/UDP with large datagrams.

**Comment by yrro** [comment]
I agree that the network is non-sensible BTW. But nonetheless it worked perfectly well for the vast majority of users who were on Windows (which presumably manages to do EDNS/UDP with large packets), hence no one noticed that DNS/TCP was completely broken for years before I found that my Linux machine didn't work properly on the network. ;)
